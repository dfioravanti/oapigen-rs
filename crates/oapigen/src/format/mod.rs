pub(crate) mod numbers;
pub(crate) mod strings;
pub use numbers::format_number;
pub use strings::format_string;

pub const DEFAULT_NUMBER: &str = "float";
pub const DEFAULT_INTEGER: &str = "integer";
pub const DEFAULT_STRING: &str = "string";
