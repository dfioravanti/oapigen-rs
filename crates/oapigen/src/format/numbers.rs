//! [https://spec.openapis.org/registry/format/]

use crate::models;
use log::warn;
use proc_macro2::TokenStream;
use quote::quote;
use std::collections::HashMap;

/// Formats numbers (including integers) according to the registry provided in
/// [https://spec.openapis.org/registry/format/].
///
/// The list of supported format is a subset of the registry.
pub fn format_number(type_format: &str) -> (TokenStream, models::Imports) {
    let tokenized_type: TokenStream;
    let imports = HashMap::new();

    match type_format {
        "int64" => tokenized_type = quote! { i64 },
        "integer" | "int32" => tokenized_type = quote! { i32 },
        "int16" => tokenized_type = quote! { i16 },
        "int8" => tokenized_type = quote! { i8 },
        "uint64" => tokenized_type = quote! { u64 },
        "uint32" => tokenized_type = quote! { u32 },
        "uint16" => tokenized_type = quote! { u16 },
        "uint8" => tokenized_type = quote! { u8 },
        "float" | "number " => tokenized_type = quote! { f32 },
        "double" => tokenized_type = quote! { f64 },
        _ => {
            tokenized_type = {
                warn!("format {type_format} is unknown for numbers, defaulting to number");
                quote! { f32 }
            }
        }
    }

    (tokenized_type, imports)
}
