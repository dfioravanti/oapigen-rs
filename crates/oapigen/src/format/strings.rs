//! [https://spec.openapis.org/registry/format/]

use crate::models;
use crate::models::DateTimeLibraries;
use log::warn;
use proc_macro2::TokenStream;
use quote::quote;
use std::collections::HashMap;

// the various names for the imports unsed in the various
const NAME_IMPORTS_CHRONO: &str = "chrono_datetime_utc";
const NAME_IMPORTS_JIFF: &str = "jiff_timestamp";

pub(crate) fn default_string() -> (TokenStream, models::Imports) {
    (quote! { String }, HashMap::new())
}

/// Formats strings according to the registry provided in
/// [https://spec.openapis.org/registry/format/].
///
/// The list of supported format is a subset of the registry.
pub fn format_string(config: &models::Config, type_format: &str) -> (TokenStream, models::Imports) {
    let tokenized_type: TokenStream;
    let mut imports = HashMap::new();

    match type_format {
        "date-time" => match config.libraries.datetime {
            DateTimeLibraries::Chrono => {
                tokenized_type = quote! { DateTime<Utc> };
                imports.insert(
                    NAME_IMPORTS_CHRONO.to_string(),
                    quote! { use chrono::{DateTime, Utc}; },
                );
            }
            DateTimeLibraries::Jiff => {
                tokenized_type = quote! { Timestamp };
                imports.insert(
                    NAME_IMPORTS_JIFF.to_string(),
                    quote! { use jiff::Timestamp; },
                );
            }
        },
        _ => {
            warn!("format {type_format} is unknown for strings, defaulting to string");
            return default_string();
        }
    }

    (tokenized_type, imports)
}
