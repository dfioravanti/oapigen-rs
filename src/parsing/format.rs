//! [https://spec.openapis.org/registry/format/]

use log::warn;
use proc_macro2::TokenStream;
use quote::quote;

pub fn type_for_format_string(type_format: &str) -> TokenStream {
    match type_format {
        "time" => {
            todo!()
        }
        _ => {
            warn!("format {type_format} is unknown for strings, defaulting to string");
            quote! { String }
        }
    }
}
