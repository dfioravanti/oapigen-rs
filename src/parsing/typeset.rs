use crate::parsing::format::type_for_format_string;
use log::warn;
use oas3::spec::{ObjectSchema, Schema, SchemaType, SchemaTypeSet};
use proc_macro2::TokenStream;
use quote::{ToTokens, quote};

pub fn tokenize_schema2(schema_name: String, schema: ObjectSchema) -> TokenStream {
    match &schema.schema_type {
        Some(schema_typeset) => match schema_typeset {
            SchemaTypeSet::Single(schema_type) => {
                tokenize_schema_type(schema_name, &schema, schema_type)
            }
            SchemaTypeSet::Multiple(items) => todo!(),
        },
        None => todo!(),
    }
}

fn tokenize_schema_type(
    schema_name: String,
    schema: &ObjectSchema,
    schema_type: &SchemaType,
) -> TokenStream {
    match schema_type {
        SchemaType::Boolean => todo!(),
        SchemaType::Integer => tokenize_integer_schema(schema_name, schema),
        SchemaType::Number => tokenize_number_schema(schema_name, schema),
        SchemaType::String => tokenize_string_schema(schema_name, schema),
        SchemaType::Array => todo!(),
        SchemaType::Object => todo!(),
        SchemaType::Null => todo!(),
    }
}

/// [https://spec.openapis.org/registry/format/]
fn tokenize_integer_schema(schema_name: String, schema: &ObjectSchema) -> TokenStream {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let tokenized_type = match &schema.format {
        Some(format) => todo!(),
        None => quote! { i64 },
    };

    quote! {
        #tokenized_name : #tokenized_type
    }
}

fn tokenize_number_schema(schema_name: String, schema: &ObjectSchema) -> TokenStream {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let tokenized_type = match &schema.format {
        Some(format) => todo!(),
        None => quote! { float },
    };

    quote! {
        #tokenized_name : #tokenized_type
    }
}

fn tokenize_string_schema(schema_name: String, schema: &ObjectSchema) -> TokenStream {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse().unwrap();

    let tokenized_type = match &schema.format {
        Some(format) => type_for_format_string(format),
        None => quote! { String },
    };

    quote! {
        #tokenized_name : #tokenized_type
    }
}

// This module deals with parsing of the various formats supported by this library.
// The list of supported formats is a subset of
// [https://spec.openapis.org/registry/format/]
pub fn parse_format(schema_type: SchemaType, format: String) -> TokenStream {
    match format {
        _ => todo!(), // _ => {
                      //     warn!(
                      //         "unknown format {:} for type {:?}, using the default for {:?}",
                      //         format, schema_type, schema_type
                      //     );
                      // }
    }
}

#[cfg(test)]
mod tests {
    use test_case::test_case;

    use super::*;

    #[test_case("age", "type: integer", "age : i64"; "integer")]
    #[test_case("age", "type: number", "age : float"; "number")]
    #[test_case("age", "type: string", "age : String"; "string")]
    fn test_parse_base_cases(schema_name: &str, schema_spec: &str, expected: &str) {
        let schema = serde_yaml::from_str::<ObjectSchema>(schema_spec).unwrap();

        let got = tokenize_schema2(schema_name.to_string(), schema);

        assert_eq!(got.to_string(), expected.to_string());
    }
}
