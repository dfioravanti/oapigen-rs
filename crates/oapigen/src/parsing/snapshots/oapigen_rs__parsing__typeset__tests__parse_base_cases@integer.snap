---
source: src/parsing/typeset.rs
expression: got.to_string()
---
type Age = i32 ;
